decoder/decoder_test.go

```diff
@@ -41,6 +41,10 @@ func testPathDecoder(t *testing.T, pathCtx *PathContext) *PathDecoder {
 	d := NewDecoder(&testPathReader{
 		paths: dirs,
 	})
+	d.SetContext(DecoderContext{
```

Comment:
A common convention is that if the struct isn't safe to use directly (i.e. one has to allocate maps or slices because the underlying code doesn't do it lazily), it has a constructor, such as `NewDecoderContext()`.

So I think we should either do the lazy allocation, and/or check for `nil`s where we may be accessing unallocated maps, or create and use the constructor.